File: system-tests/maven-external-upload-test/src/test/java/com/teamscale/upload/MavenExternalUploadSystemTest.java
package com.teamscale.upload;

import com.teamscale.test.commons.ExternalReport;
import com.teamscale.test.commons.SystemTestUtils;
import com.teamscale.test.commons.TeamscaleMockServer;
import org.apache.commons.lang3.SystemUtils;
import org.conqat.lib.commons.filesystem.FileSystemUtils;
import org.conqat.lib.commons.io.ProcessUtils;
import org.junit.jupiter.api.AfterAll;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.io.TempDir;

import java.io.File;
import java.io.IOException;
import java.nio.file.Path;
import java.nio.file.Paths;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.fail;

/**
 * Runs several Maven projects' Surefire and Failsafe tests that produce coverage via the Jacoco Maven plugin. Checks
 * that the Jacoco reports are correctly uploaded to a Teamscale instance.
 */
public class MavenExternalUploadSystemTest {

	private static final String MAVEN_COVERAGE_UPLOAD_GOAL = "com.teamscale:teamscale-maven-plugin:upload-coverage";

	private static TeamscaleMockServer teamscaleMockServer = null;

	private static final String NESTED_MAVEN_PROJECT_NAME = "nested-project";

	private static final String FAILING_MAVEN_PROJECT_NAME = "failing-project";

	@BeforeEach
	public void startFakeTeamscaleServer() throws Exception {
		if (teamscaleMockServer == null) {
			teamscaleMockServer = new TeamscaleMockServer(SystemTestUtils.TEAMSCALE_PORT).acceptingReportUploads();
		}
		teamscaleMockServer.uploadedReports.clear();
	}

	private ProcessUtils.ExecutionResult runCoverageUploadGoal(String projectPath) {
		File workingDirectory = new File(projectPath);
		String executable = "./mvnw";
		if (SystemUtils.IS_OS_WINDOWS) {
			executable = Paths.get(projectPath, "mvnw.cmd").toUri().getPath();
		}
		try {
			return ProcessUtils.execute(
					new ProcessBuilder(executable, MAVEN_COVERAGE_UPLOAD_GOAL).directory(workingDirectory));
		} catch (IOException e) {
			fail(String.valueOf(e));
		}
		return null;
	}

	@Test
	public void testMavenExternalUpload() throws Exception {
		SystemTestUtils.runMavenTests(NESTED_MAVEN_PROJECT_NAME);
		runCoverageUploadGoal(NESTED_MAVEN_PROJECT_NAME);
		assertThat(teamscaleMockServer.uploadedReports).hasSize(6);
		ExternalReport unitTests = teamscaleMockServer.uploadedReports.get(0);
		ExternalReport integrationTests = teamscaleMockServer.uploadedReports.get(3);
		assertThat(unitTests.getPartition()).isEqualTo("My Custom Unit Tests Partition");
		assertThat(integrationTests.getPartition()).isEqualTo("Integration Tests");
	}

	@Test
	public void testIncorrectJaCoCoConfiguration() throws IOException {
		SystemTestUtils.runMavenTests(FAILING_MAVEN_PROJECT_NAME);
		ProcessUtils.ExecutionResult result = runCoverageUploadGoal(FAILING_MAVEN_PROJECT_NAME);
		assertThat(result).isNotNull();
		assertThat(teamscaleMockServer.uploadedReports).isEmpty();
		assertThat(result.getStdout()).contains(
				String.format("Skipping upload for %s as %s is not configured to produce XML reports",
						FAILING_MAVEN_PROJECT_NAME, "org.jacoco:jacoco-maven-plugin"));
	}

	/**
	 * When no commit is given and no git repo is available, which is the usual fallback, a helpful error message should
	 * be shown (TS-40425).
	 */
	@Test
	public void testErrorMessageOnMissingCommit(@TempDir Path tmpDir) throws IOException {
		FileSystemUtils.copyFiles(new File("missing-commit-project"), tmpDir.toFile(), file -> true);
		tmpDir.resolve("mvnw").toFile().setExecutable(true);
		String projectPath = tmpDir.toAbsolutePath().toString();
		SystemTestUtils.runMavenTests(projectPath);
		ProcessUtils.ExecutionResult result = runCoverageUploadGoal(projectPath);
		assertThat(result).isNotNull();
		assertThat(result.getReturnCode()).isNotEqualTo(0);
		assertThat(teamscaleMockServer.uploadedReports).isEmpty();
		assertThat(result.getStdout()).contains("There is no <commit> configured in the pom.xml and it was not possible to determine the checked out commit");
	}

	@AfterAll
	public static void stopFakeTeamscaleServer() {
		teamscaleMockServer.shutdown();
	}

}

File: system-tests/maven-external-upload-test/src/test/kotlin/com/teamscale/upload/MavenExternalUploadSystemTest.kt
package com.teamscale.upload

import com.teamscale.client.EReportFormat
import com.teamscale.test.commons.ExternalReport
import com.teamscale.test.commons.Session
import com.teamscale.test.commons.SystemTestUtils
import com.teamscale.test.commons.SystemTestUtils.runMavenTests
import com.teamscale.test.commons.TeamscaleMockServer
import org.apache.commons.lang3.SystemUtils
import org.assertj.core.api.Assertions
import org.assertj.core.api.Assertions.assertThat
import org.conqat.lib.commons.filesystem.FileSystemUtils
import org.conqat.lib.commons.io.ProcessUtils
import org.junit.jupiter.api.AfterAll
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.io.TempDir
import java.io.File
import java.io.FileFilter
import java.io.IOException
import java.nio.file.Path
import java.nio.file.Paths

/**
 * Runs several Maven projects' Surefire and Failsafe tests that produce coverage via the Jacoco Maven plugin. Checks
 * that the Jacoco reports are correctly uploaded to a Teamscale instance.
 */
class MavenExternalUploadSystemTest {
	@BeforeEach
	@Throws(Exception::class)
	fun startFakeTeamscaleServer() {
		if (teamscaleMockServer == null) {
			teamscaleMockServer = TeamscaleMockServer(SystemTestUtils.TEAMSCALE_PORT).acceptingReportUploads()
		}
		teamscaleMockServer?.reset()
	}

	private fun runCoverageUploadGoal(projectPath: String): ProcessUtils.ExecutionResult? {
		val workingDirectory = File(projectPath)
		var executable = "./mvnw"
		if (SystemUtils.IS_OS_WINDOWS) {
			executable = Paths.get(projectPath, "mvnw.cmd").toUri().getPath()
		}
		try {
			return ProcessUtils.execute(
				ProcessBuilder(executable, MAVEN_COVERAGE_UPLOAD_GOAL).directory(workingDirectory)
			)
		} catch (e: IOException) {
			Assertions.fail(e.toString())
		}
		return null
	}

	@Test
	@Throws(Exception::class)
	fun testMavenExternalUpload() {
		runMavenTests(NESTED_MAVEN_PROJECT_NAME)
		runCoverageUploadGoal(NESTED_MAVEN_PROJECT_NAME)
		val sessions = teamscaleMockServer!!.getSessions()
		assertThat(sessions).hasSize(2)
		assertThat(sessions.first().partition).isEqualTo("Integration Tests")
		assertThat(sessions.last().partition).isEqualTo("My Custom Unit Tests Partition")

		assertThat(sessions.first().getReports(EReportFormat.JACOCO)).hasSize(3)
		assertThat(sessions.last().getReports(EReportFormat.JACOCO)).hasSize(3)
	}

	@Test
	@Throws(IOException::class)
	fun testIncorrectJaCoCoConfiguration() {
		runMavenTests(FAILING_MAVEN_PROJECT_NAME)
		val result = runCoverageUploadGoal(FAILING_MAVEN_PROJECT_NAME)
		assertThat(result).isNotNull()
		assertThat(teamscaleMockServer!!.getSessions()).isEmpty()
		assertThat(result!!.stdout).contains(
			"Skipping upload for $FAILING_MAVEN_PROJECT_NAME as org.jacoco:jacoco-maven-plugin is not configured to produce XML reports"
		)
	}

	/**
	 * When no commit is given and no git repo is available, which is the usual fallback, a helpful error message should
	 * be shown (TS-40425).
	 */
	@Test
	@Throws(IOException::class)
	fun testErrorMessageOnMissingCommit(@TempDir tmpDir: Path) {
		FileSystemUtils.copyFiles(File("missing-commit-project"), tmpDir.toFile()) { true }
		tmpDir.resolve("mvnw").toFile().setExecutable(true)
		val projectPath = tmpDir.toAbsolutePath().toString()
		runMavenTests(projectPath)
		val result = runCoverageUploadGoal(projectPath)
		assertThat(result).isNotNull()
		assertThat(result!!.returnCode).isNotEqualTo(0)
		assertThat(teamscaleMockServer!!.getSessions()).isEmpty()
		assertThat(result.stdout)
			.contains("There is no <revision> or <commit> configured in the pom.xml and it was not possible to determine the current revision")
	}

	companion object {
		private var teamscaleMockServer: TeamscaleMockServer? = null

		private const val MAVEN_COVERAGE_UPLOAD_GOAL = "com.teamscale:teamscale-maven-plugin:upload-coverage"
		private const val NESTED_MAVEN_PROJECT_NAME = "nested-project"
		private const val FAILING_MAVEN_PROJECT_NAME = "failing-project"

		@JvmStatic
		@AfterAll
		fun stopFakeTeamscaleServer() {
			teamscaleMockServer?.shutdown()
		}
	}
}
